src/genai/extensions/llama_index/llm.py

```diff
@@ -0,0 +1,128 @@
+import asyncio
+import logging
+from functools import partial
+from typing import Any, List, Optional, Sequence
+
+from genai import Model
+from genai.schemas.chat import (
+    AIMessage,
+    BaseMessage,
+    ChatRole,
+    HumanMessage,
+    SystemMessage,
+)
+from genai.schemas.responses import ModelCard
+
+logger = logging.getLogger(__name__)
+
+
+try:
+    from llama_index.callbacks import CallbackManager
+    from llama_index.llms.base import (
+        LLM,
+        ChatMessage,
+        ChatResponse,
+        ChatResponseAsyncGen,
+        ChatResponseGen,
+        CompletionResponse,
+        CompletionResponseAsyncGen,
+        CompletionResponseGen,
+        LLMMetadata,
+        MessageRole,
+        llm_chat_callback,
+        llm_completion_callback,
+    )
+
+    class LlamaIndexChatMessageClasses:
+        user = ChatMessage
+        system = ChatMessage
+        assistant = ChatMessage
+
+except ImportError:
+    raise ImportError("Could not import llamaindex: Please install ibm-generative-ai[llama-index] extension.")
+
+
+def to_genai_message(message: ChatMessage) -> BaseMessage:
+    if message.role == ChatRole.user:
+        return HumanMessage(content=message.content)
+    elif message.role == ChatRole.system:
+        return SystemMessage(content=message.content)
+    elif message.role == ChatRole.assistant:
+        return AIMessage(content=message.content)
+    else:
+        raise ValueError(f"Got unknown message type {message}")
+
+
+def to_genai_messages(messages: List[ChatMessage]) -> List[BaseMessage]:
+    return [to_genai_message(msg) for msg in messages]
+
+
+class IBMGenAILlamaIndex(LLM):
+    model: Model
+    model_info: Optional[ModelCard]
+
+    def __init__(self, model: Model, callback_manager: Optional[CallbackManager] = None, **kwargs: Any):
+        super().__init__(model=model, callback_manager=callback_manager, **kwargs)
+
+        self.model_info = self.model.info()
+
+    @classmethod
+    def class_name(cls) -> str:
+        "ibmgenai_llm"
+
+    @property
+    def metadata(self) -> LLMMetadata:
+        return LLMMetadata(
+            context_window=self.model_info.token_limit,
+            is_chat_model=True,
+            is_function_calling_model=False,
+            model_name=self.model_info.name,
+        )
+
+    @llm_completion_callback()
+    def complete(self, prompt: str, **kwargs: Any) -> CompletionResponse:
```

Comment:
Instead of plain `**kwargs`, add some parameters which are used in `self.model.generate` because I suppose this method is publicly accessible to the user.

Reply:
this should be preserved everywhere (method below)